mechanics/physicell/tests/test_vtk_agents_serializer.cpp

```diff
@@ -0,0 +1,303 @@
+#include <array>
+#include <filesystem>
+#include <fstream>
+#include <iterator>
+#include <memory>
+#include <vector>
+#include <vtkPointData.h>
+#include <vtkSmartPointer.h>
+#include <vtkUnstructuredGrid.h>
+#include <vtkXMLUnstructuredGridReader.h>
+
+#include <gtest/gtest.h>
+
+#include "physicell/mechanical_agent_container.h"
+#include "physicell/vtk_agents_serializer.h"
+
+using namespace physicore;
+using namespace physicore::mechanics::physicell;
+
+namespace {
+
+mechanical_agent_container make_container(int dims, int agent_types, int substrates)
+{
+	auto base = std::make_unique<base_agent_data>(dims);
+	auto data = std::make_unique<agent_data>(*base, agent_types, substrates);
+	return mechanical_agent_container(std::move(base), std::move(data));
+}
+
+} // namespace
+
+class VtkMechanicsAgentsSerializerTest : public ::testing::Test
+{
+protected:
+	std::filesystem::path test_output_dir;
+
+	void SetUp() override
+	{
+		test_output_dir = std::filesystem::temp_directory_path() / "vtk_mechanics_agents_serializer_test";
+		std::filesystem::remove_all(test_output_dir);
+		std::filesystem::create_directories(test_output_dir);
+	}
+
+	void TearDown() override
+	{
+		if (std::filesystem::exists(test_output_dir))
+		{
+			std::filesystem::remove_all(test_output_dir);
+		}
+	}
+};
+
+TEST_F(VtkMechanicsAgentsSerializerTest, ConstructorCreatesOutputDirectories)
+{
+	auto container = make_container(3, 1, 1);
+
+	EXPECT_NO_THROW({ vtk_agents_serializer serializer(test_output_dir.string(), container); });
+
+	auto vtk_dir = test_output_dir / "vtk_mechanics_agents";
+	EXPECT_TRUE(std::filesystem::exists(vtk_dir));
+}
+
+TEST_F(VtkMechanicsAgentsSerializerTest, SerializeWithNoAgentsCreatesFiles)
+{
+	auto container = make_container(3, 1, 1);
+	vtk_agents_serializer serializer(test_output_dir.string(), container);
+
+	EXPECT_NO_THROW(serializer.serialize(0.0));
+
+	auto vtk_dir = test_output_dir / "vtk_mechanics_agents";
+	EXPECT_TRUE(std::filesystem::exists(vtk_dir / "mechanics_agents_000000.vtu"));
+	EXPECT_TRUE(std::filesystem::exists(test_output_dir / "mechanics_agents.pvd"));
+}
+
+TEST_F(VtkMechanicsAgentsSerializerTest, SerializeSingleAgentWritesExpectedArrays)
+{
+	auto container = make_container(3, 2, 2);
+	container.create();
+
+	auto& base_data = *std::get<0>(container.agent_datas);
+	auto& mech_data = *std::get<1>(container.agent_datas);
+
+	base_data.positions = { 1.0, 2.0, 3.0 };
```

Comment:
wow does this even work? you are directly modifying data structs without calling add() so agents_count is stale un-synced with the object. isnt serializer relying on agents_count?